Mathlib/Order/BooleanSubalgebra.lean

```diff
@@ -361,6 +361,26 @@ lemma closure_bot_sup_induction {p : ∀ g ∈ closure s, Prop} (mem : ∀ x hx,
       compl_mem' := fun ⟨_, hb⟩ ↦ ⟨_, compl _ _ hb⟩ }
   closure_le (L := L).mpr (fun y hy ↦ ⟨subset_closure hy, mem y hy⟩) hx |>.elim fun _ ↦ id
 
+theorem mem_closure_iff_sup_sdiff (sup : SupClosed s) (inf : InfClosed s)
+    (bot_mem : ⊥ ∈ s) (top_mem : ⊤ ∈ s) (a : α) :
+    a ∈ closure s ↔ ∃ t : Finset (s × s), a = t.sup fun x ↦ x.1.1 \ x.2.1 := by
```

Comment:
I added [an induction principle](https://github.com/leanprover-community/mathlib4/pull/20251/files#diff-a46200dbd79beae0d9c1e6367c85e13cdf48877d9af2f698bd16fe33e4004541R388-R398) that is hopefully close to what you want. I tried to prove it directly, but I don't think the last chunk of the proof (about taking complements) works without  writing the involved element explicitly as a Finset.sup of the given form.

I'm not sure about the exact form of the induction principle you want in [the Zulip thread](https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/Disjunction.20Normal.20Form.20for.20a.20generated.20boolean.20subalgebra/near/489911341); do you want to take the basis to consist of (retrocompact?) affine opens? But you may need to take infinite unions to get all retrocompact opens, right?

Reply:
> But you may need to take infinite unions to get all retrocompact opens, right?

I should not, because I am assuming the space is compact.

> Also, I think [...] in the thread should be [...]

Well spotted!